File: projective_point.py
from finite_field import F


def ProjectivePoint(a, b, mod):
    class ProjectivePoint:
        def __init__(self, x, y, z):
            self.field = F(mod)
            if isinstance(x, int):
                self.x = self.field(x)
            else:
                self.x = x
            if isinstance(y, int):
                self.y = self.field(y)
            else:
                self.y = y
            if isinstance(z, int):
                self.z = self.field(z)
            else:
                self.z = z

        def is_zero(self):
            return self.z is None

        def __add__(self, other):
            if self.is_zero():
                return other
            elif other.is_zero():
                return self
            s_1 = self.y * other.z
            s_2 = other.y * self.z
            u_1 = self.x * other.z
            u_2 = other.x * self.z
            if u_1 == u_2:
                if s_1 == s_2:
                    return self.double()
                else:
                    return ProjectivePoint(None, 1, None)
            else:
                w = self.z * other.z
                r = s_2 - s_1
                p = u_2 - u_1
                p_sqr = p * p
                rrw = r * r * w
                p_sqr_u1_u2 = p_sqr * (u_1 + u_2)
                p_cube = p * p_sqr
                rx = p * (rrw - p_sqr_u1_u2)
                ry = r * (self.field(-2)*rrw + self.field(3) * p_sqr_u1_u2) - p_cube * (s_1 + s_2)
                rz = p_cube * w
                return ProjectivePoint(self.field(2)*rx, ry, self.field(2)*rz)

        def double(self):
            if self.is_zero() or self.y == self.field(0):
                return ProjectivePoint(None, 1, None)
            else:
                two = self.field(2)
                w = self.x * self.x * self.field(3) + self.field(a) * self.z * self.z
                s = self.y * self.z * two
                b = s * self.x * self.y * two
                h = w * w - b * two
                rx = h * s  
                ry = w * (b - h) - s * s * self.y * self.y * two
                rz = s * s * s
                return ProjectivePoint(rx, ry, rz)

        def __neg__(self):
            if self.is_zero():
                return self
            else:
                return ProjectivePoint(self.x, -self.y, self.z)

        def __sub__(self, other):
            return self + -other

        def __mul__(self, n):
            if n < 0:
                return -self * -n
            result = ProjectivePoint(None, 1, None)
            temp = self
            while n != 0:
                if n & 1 != 0:
                    result += temp
                temp = temp.double()
                n >>= 1
            return result

        __rmul__ = __mul__

        def __eq__(self, other):
            if self.is_zero() or other.is_zero():
                return self.is_zero() and other.is_zero()
            else:
                return (self.x * other.z, self.y * other.z) == (other.x * self.z, other.y * self.z)

        def __ne__(self, other):
            return not (self == other)

        def __str__(self):
            if self.is_zero():
                return "(Zero)"
            else:
                return f"({self.x}, {self.y}, {self.z})"

        __repr__ = __str__

        def norm(self):
            if self.is_zero():
                return self
            else:
                return ProjectivePoint(self.x / self.z, self.y / self.z, self.field(1))

        @staticmethod
        def get_inf():
            return ProjectivePoint(None, 1, None)

    return ProjectivePoint

File: cpp/projective_point.cpp
// #include "field.cpp"
#include <iostream>




template<mpz_t a, mpz_t b, mpz_t mod>
class ProjectivePoint
{
private:
    typedef Field<mod> Number;
    Number x;
    Number y;
    Number z;
public:
    ProjectivePoint(mpz_t x, mpz_t y, mpz_t z){
       this->x = Number(x);
       this->y = Number(y);
       this->z = Number(z);
    }
    ProjectivePoint(mpz_class x, mpz_class y, mpz_class z){
       this->x = Number(x);
       this->y = Number(y);
       this->z = Number(z);
    }
    ProjectivePoint(long long x, long long y, long long z){
       this->x = Number(x);
       this->y = Number(y);
       this->z = Number(z);
    }
    ProjectivePoint(int x, int y, int z){
       this->x = Number(x);
       this->y = Number(y);
       this->z = Number(z);
    }
    ProjectivePoint(Number x, Number y ,Number z){
       this->x = x;
       this->y = y;
       this->z = z;
    }

    bool is_zero(){
        return z == Number(0);
    }

    Number getX(){
        return x;
    }

    Number getY(){
        return y;
    }

    Number getZ(){
        return z;
    }

    ProjectivePoint dbl() {
        if (this->is_zero() || this->y == Number(1)){
            return ProjectivePoint(0, 0, 0);
        }
        else{
            Number two = Number(2);
            Number w = x * x * Number(3) + Number(a) * z * z;
            Number s = y * z * two;
            Number B = s * x * y * two;
            Number h = w * w - B * two;
            Number rx = h * s ;
            Number ry = w * (B - h) - s * s * y * y * two;
            Number rz = s * s * s;
            return ProjectivePoint(rx, ry, rz);
        }
    }

    ProjectivePoint operator + (ProjectivePoint const &other) {
        if(this->is_zero()) {
            return other;
        }
        else {
            if (Number(other.z) == Number(0)) {
                return *this;
            }
        }
        Number s_1 = this->y * other.z;
        Number s_2 = this->z * other.y;
        Number u_1 = this->x * other.z;
        Number u_2 = this->z * other.x;
        if (u_1 == u_2) {
            if (s_1 == s_2) {
                return this->dbl();
            } else {
                return ProjectivePoint(0, 1, 0);
            }
        } else {
            Number w = z * other.z;
            Number r = s_2 - s_1; 
            Number p = u_2 - u_1;   
            Number p_sqr = p * p;
            Number rrw = r * r * w;
            Number p_sqr_u1_u2 = p_sqr * (u_1 + u_2);
            Number p_cube = p * p_sqr;
            Number rx = p * (rrw - p_sqr_u1_u2);
            Number ry = r * (Number(-2)*rrw + Number(3) * p_sqr_u1_u2) - p_cube * (s_1 + s_2);
            Number rz = p_cube * w;
            return ProjectivePoint(Number(2) * rx, ry, Number(2) * rz);
        }
    }

    ProjectivePoint operator - (ProjectivePoint const &other) {
        return *this + -other;
    }

    ProjectivePoint operator - () {
        if (this->is_zero()) {
            return *this;
        } else {
            return ProjectivePoint(this->x, -this->y, this->z);
        }
    }

    ProjectivePoint operator * (mpz_class n){
        if (n < 0) {
            return -(*this) * -n;
        }
        ProjectivePoint result = ProjectivePoint(0, 0, 0);
        ProjectivePoint temp = *this;
        while(n != 0){
            if (n % 2 != 0){
                result = result + temp;
            }
            temp =  temp.dbl();
            n = n >> 1;
        }
        return result;
    }

    bool operator != (ProjectivePoint const &other) {
        return !(this == other);
    }

    bool operator == (ProjectivePoint const &other){
         if(this->is_zero() || Number(other.x) == Number(0)){
             return this->is_zero() && Number(other.x) == Number(0);
         }
         else {
             return (this->x * other.z == this->z * other.x ) && (this->y * other.z == this->z * other.y );
         }
    }

    friend std::ostream & operator << (std::ostream &out, ProjectivePoint &point){
        out << "(" << point.x.getValue()<< ", " << point.y.getValue() << ", " << point.z.getValue() <<")";
        return out;
    }

};

